query/src/sql/planner/bind_context.rs

```diff
@@ -0,0 +1,163 @@
+// Copyright 2021 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use common_datavalues::DataType;
+use common_exception::ErrorCode;
+use common_exception::Result;
+
+use crate::sql::common::IndexType;
+use crate::sql::opt::SExpr;
+use crate::sql::parser::ast::TableAlias;
+use crate::sql::planner::scalar::ScalarExpr;
+
+#[derive(Clone)]
+pub struct ColumnBinding {
+    /// Table name of this `ColumnBinding` in current context
+    pub table_name: String,
+    /// Column name of this `ColumnBinding` in current context
+    pub column_name: String,
+    pub index: IndexType,
+    pub data_type: DataType,
+    pub nullable: bool,
+    pub expr: Option<ScalarExpr>,
+}
+
+/// `BindContext` stores all the free variables in a query and tracks the context of binding procedure.
+#[derive(Clone, Default)]
+pub struct BindContext {
+    _parent: Option<Box<BindContext>>,
```

Comment:
I guess this member is reserved for correlated subqueries?

Reply:
Yes, we need to resolve outer columns with a stack-like structure.

Reply:
a Box<BindContext> means this BindContext owns this _parent, if the _parent a reference to a parent BindContext object, it'd be a weak reference pointer?

Reply:
> a Box means this BindContext owns this _parent, if the _parent a reference to a parent BindContext object, it'd be a weak reference pointer?

It's complicated to maintain a recursive data structure in rust, so I just make parent an owned element and won't be shared.